### PR TITLE
Update the relay specific specifications to consider them generic best practices and not purely a relay spec

### PR DESCRIPTION
**Pull request by orta**
Better late than never, roughly a year ago in the GraphQL Working Group 3 we took it on Artsy to start moving the Relay specs to be considered "GraphQL Best Practices" ([link](https://github.com/graphql/graphql-wg/blob/master/notes/2018-02-01.md#discuss-moving-connectionglobal-id-specs-from-relay-into-best-practices-repo))

This starts that, beginning with first minimizing the use of Relay as being the sole client that conforms to the connection and object identifier spec. 

Main notes:

- When referring to relay on the client I switched "Relay" to "Spec-compliant client" or "client"
- When referring to relay's compatibility on the server side, I oriented it to just be "spec-compliant"
- I mostly left the Mutation input spec alone, now that it's only for Relay classic and is close to being removed completely from Relay https://github.com/facebook/relay/pull/2349
- This is step one, once this is merged then I'd like to move both object identification and connection to be on the graphql repos instead of relay and I can handle link updating etc

Happy to change wording etc

/cc @leebyron in case you don't watch this repo too closely.

### CODE REVIEW

File: website/spec/Mutations.md [hunk 1]
```diff
@@ -1,9 +1,9 @@
-Relay Input Object Mutations Specification
+Relay Classic Input Object Mutations Specification
 ------------------------------------------
 
-Relay's support for mutations relies on the GraphQL server exposing
+Relay Classic's support for mutations relies on the GraphQL server exposing
 mutation fields in a standardized way. These mutations accept and emit a
-identifier string, which allows Relay to track mutations and responses.
+identifier string, which allows Relay Classic to track mutations and responses.
```
[the remark this answers is not in the document]
Comment (orta) [reply website/spec/Mutations.md]: Is it used though? If the functionality still existing in the codebase but isn't a part of how you have to do mutations then it should be fine

[the remark this answers is not in the document]
Comment (robrichard) [reply website/spec/Mutations.md]: There's nothing in Relay Modern that expects `clientMutationId` to be there. It's just one possible workaround to fix the issue described here: https://github.com/facebook/relay/issues/1734#issuecomment-304782231

[the remark this answers is not in the document]
Comment (alloy) [reply website/spec/Mutations.md]: Ahh ok, I misremembered. Cool, yeah then it can stay as-is 👍

**Comment by josephsavona** [comment]
@alloy that is closed though - is there any actual GraphQL docs that mention these as best practices?

**Comment by orta** [comment]
No, there aren't any docs that specify that - as of right now, these Specs are totally considered relay specific. 

The term best practices could be switched with "GraphQL Schema Extensions" or something among those lines. However, that best practices was the term we used in the meeting a year ago, but can definitely change. The idea being that they are not a part of the spec, but represent a consolidated approach to problems that can/should be adopted in most schemas.

**Comment by josephsavona** [comment]
Yeah, I agree that these *should* be considered best practices, but it would be great if we could have similar messaging on GraphQL docs themselves. Hmm.

**Comment by leebyron** [comment]
This looks great to me as a first step.

I'd actually suggest the terminology "Common Patterns" instead of "Best Practices" since what we're trying to do with docs like these is capture common problems people encounter when building GraphQL APIs and the best working patterns to apply to those problems.

This isn't really an extension of the schema or spec since these all fit well within the existing spec, they aren't suggesting any additional extensions beyond what GraphQL defines. GraphQL spec just defines the mechanisms by which it works, and the spec itself isn't a great place for trying to codify API patterns like this. However I think the official graphql website docs are a great place for that

As a next step, I would suggest opening a PR on the graphql.org website (https://github.com/graphql/graphql.github.io) which updates https://graphql.org/learn/ to contain a new third header on the side bar "Common Patterns" with a document for each of these api patterns.

The graphql site also uses markdown for docs, so hopefully porting them over is pretty easy to do.

Then, I think it would be helpful to replace the docs on the relay site with a much shorter statement that, if you use these common patterns, Relay leverages them to enable more sophisticated behavior like automatic pagination and smarter caching

**Comment by orta** [comment]
Perfect, thanks! I'll get on with this, agree on terminology

**Comment by josephsavona** [comment]
I disagree that "common patterns" is appropriate terminology for pagination: "best practice" is the standard terminology to describe a recommended approach to solving a particular problem. The Relay connection spec is *the* recommended way to implement pagination in GraphQL - per graphql.org, which already includes an informal description of it [*in the best practices section*](https://graphql.org/learn/pagination/). We should make update that section to include (or link to) the more formal version of the connection spec, and then link to it from the Relay docs as a best practice. 

> these all fit well within the existing spec

Object identity is a different case  - it's a fundamental part of the problem domain that should be covered by the GraphQL spec; the intentional exclusion of this from the spec has led to serious fragmentation in the ecosystem. Relay/Facebook's `Node` interface is a reasonable workaround to this deficiency in the spec, but other approaches are equally reasonable. Including Relay's approach to object identity as a "common pattern" on graphql.org/learn seems like a good compromise.

So to clarify, the steps are:
* Send a PR to graphql.org to change https://graphql.org/learn/pagination/ to include or link to the formal version of Relay's connection spec (with naming changed to "GraphQL Connection Spec" or similar). Then update Relay docs to remove the local version and link to the graphql.org one.
* Send a PR to graphql.org to add a "Common Patterns" section of the /learn page and with an informal description of Relay's approach to object identity with a link to the formal spec, w name changed to "GraphQL Object Identity Spec" or similar. Then update Relay docs to remove the local version and link to the graphql.org one.
* Longer-term: change the GraphQL spec to include an official approach to object identity.

@orta Interested in updating this PR along those lines and sending PRs to graphql.org?

**Comment by orta** [comment]
Yep this makes sense @josephsavona  👍  - if I get time tomorrow I'll move this in that direction

**Comment by orta** [comment]
Hey folks, either this week or next (this week is pretty busy for me) - I'm going to update _all_ these PRs including https://github.com/graphql/graphql.github.io/pull/665 and https://github.com/graphql/graphql.github.io/pull/666 

Then for the graphQL website side, I'm going to give a week for them to settle then merge the PRs so that from that side the relay specs are recommended as best practices

**Comment by orta** [comment]
OK, I think this is everything - all of the GraphQL website side is ready and in production. I've updated this PR and included links to the graphql pages.

Identification lives in:  https://graphql.org/learn/global-object-identification/

The Connection spec still lives on Relay, but it is now linked to from the graphQL site:  https://graphql.org/learn/pagination/#connection-specification

**Pull request closed without merging**